packages/react-utilities/src/hooks/useCallbackRef.ts

```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
+
+/**
+ * Creates a MutableRef with ref change callback. Is useful as React.useRef() doesn't notify you when its content
```

Comment:
I guess my all-up question with this hook is whether it might be encouraging anti-patterns, or be prone to misuse which introduces unexpected behavior

Reply:
Looking at the current usage kind of reinforces this concern. The hook introduces essentially a "magic ref" which looks exactly like a normal ref, except updating it causes re-renders. That seems like a bit of a "footgun." Not to say this is unresolvable or we shouldn't have this hook at all, just that we should consider ways to reduce the "footgun" potential.

Reply:
I will move this to `react-positioning` until we need it elsewhere, I agree with you on dangerous ref

Reply:
> The hook introduces essentially a "magic ref" which looks exactly like a normal ref, except updating it causes re-renders.

A small correction there: it will not cause re-renders, it will just a callback on every change 😉

But, in general, I agree that it's to magical for public API and should be used accurately.

Reply:
> "magic ref" which looks exactly like a normal ref, except updating it causes re-renders

To clarify the hook does not cause rerenders.. it calls a callback which **can** cause rerenders but most of the time you would use it to do ref operations because you want to **avoid** rerenders